### ISSUE TITLE
Assistance with proper struct setup

### ISSUE THREAD

**Issue by dfischer** [zevdg/fcf#1]
I'm having a bit of trouble figuring out the right way to setup my types to pick up this structure. 

On my server this is the natural type:

```
// Assessment stores assessment in db
type Assessment struct {
	TemplateID string               `firestore:"templateID"`
	Created    time.Time            `firestore:"created"`
	UserID     string               `firestore:"userID"`
	Slots      map[string]*bot.Slot `firestore:"slots"`
}

type Slot struct {
	Value   interface{} `firestore:"value"`
	Created time.Time   `firestore:"created"`
}
```

Which then gets sent as this to the firestore trigger cloud function:

```
{
  "oldValue": {},
  "updateMask": {},
  "value": {
    "createTime": "2019-06-13T22:44:02.081775Z",
    "fields": {
      "created": {
        "timestampValue": "2019-06-13T22:44:01.926976Z"
      },
      "slots": {
        "mapValue": {
          "fields": {
            "frustrated": {
              "mapValue": {
                "fields": {
                  "created": {
                    "timestampValue": "2019-06-13T22:43:57.332848Z"
                  },
                  "value": {
                    "stringValue": "zxc"
                  }
                }
              }
            },
            "hitQuota": {
              "mapValue": {
                "fields": {
                  "created": {
                    "timestampValue": "2019-06-13T22:43:53.571678Z"
                  },
                  "value": {
                    "stringValue": "qwe"
                  }
                }
              }
            },
            "toolsTechnology": {
              "mapValue": {
                "fields": {
                  "created": {
                    "timestampValue": "2019-06-13T22:43:55.566685Z"
                  },
                  "value": {
                    "stringValue": "asd"
                  }
                }
              }
            }
          }
        }
      },
      "templateID": {
        "stringValue": "ae"
      },
      "userID": {
        "stringValue": "1667237206656764"
      }
    },
    "name": "removed",
    "updateTime": "2019-06-13T22:44:02.081775Z"
  }
}
```

The goal is to iterate over every slot easily and grab that value. That's where I'm having the most trouble. I'm close but... feel like I may be doing it wrong and missing something from a mental model perspective.

Here's the type on the cloud function 

```
type assessmentData struct {
	TemplateID string    `fcf:"templateID"`
	Created    time.Time `fcf:"created"`
	UserID     string    `fcf:"userID"`

	MapVal map[string]struct {
		ValString map[string]string `json:"stringValue"`
	} `fcf:"slots"`
	MapDynamic   map[string]interface{} `fcf:"slots"`
	MapInterface interface{}            `fcf:"slots"`
}
```

I'm experimenting so there's the superfluous last two attributes.

Basically just want to output each value in slot as finishing my experiment.

```
	for k, v := range a.MapVal {
		log.Printf("MapVal: k: %s, val: %+v", k, v)
		log.Printf("MapVal: v : %+v", v)
		for kk, vv := range v.ValString { // this would be the slot attributes
			log.Printf("ValString: kk: %s, val: %+v", kk, vv)
			log.Printf("ValString: vv: %+v", vv)
		}
	}
```

I tried it without having the nested struct with ValString but wasn't having much luck there so trying to do this piece more manually but still close but no cigar.

Anyway, thanks!

edit:  note:

Usually I'd expect to just use a `map[string]type` structure for dynamic keys but this is proving difficult with the payload - want to simplify if possible.

**Comment by zevdg** [comment]
So I think i've recreated this issue and it does look like a bug in fcf.  It looks like i'm missing two important cases in my test suite:

* a struct inside a map: map[string]SomeType
* a struct pointer inside a map: map[string]*SomeType

Both of these ought to work, but in my preliminary tests, it looks like the first panics, and the second gives this error: Error unmarshalling field foo["bar"]: type mismatch: Cannot unmarshal firestore mapValue into a ptr field

Thanks for letting me know!
I'll try to look at this soon, but if you beat me to it, I do accept pull requests. :smiley:

**Comment by dfischer** [comment]
That's exactly the errors I was getting as well but just thought I was doing something wrong and I was pretty braindead. Haha. I unfortunately haven't dug in too heavily on reflection yet in Go but I was doing some deep dives in encoding/json internals because of this. Not sure how much time I'll have this week but I'll get to it sooner or later if you haven't done it yet. :) I totally get it!

**Comment by zevdg** [comment]
Ok, I finally got around the debugging and fixing both of these issues.  Please update to the latest release and let me know if everything is working for you.
